### DISCUSSION TITLE
Async select options not selecting defaultValue

### DISCUSSION THREAD

**Discussion by badsyntax** [react-hook-form/react-hook-form#7750] [Q&A] [answered]
Hi here. I'm having a hard time deciding on the correct approach for setting a `<select>` defaultValue when the options list is async rendered, as the rendered selected option does not match the default value.

A real-world scenario is rendering the form with empty option list, but loading the list of options via http request.

What's the best approach for handling this scenario?

Here's an example (codesandbox here: https://codesandbox.io/s/react-hook-form-get-started-forked-ljkfu?file=/src/index.js):

```jsx
import React, { useEffect, useState } from "react";
import ReactDOM from "react-dom";
import { useForm } from "react-hook-form";

import "./styles.css";

function App() {
  const { register, handleSubmit, watch } = useForm({
    defaultValues: {
      example: "2"
    }
  });
  const onSubmit = (data) => {
    console.log(data);
  };

  // Change the default value to true to see the correct select state
  const [loaded, setLoaded] = useState(false);

  useEffect(() => {
    setTimeout(() => {
      setLoaded(true);
    }, 1000);
  }, []);

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <select {...register("example")}>
        {loaded && (
          <>
            <option value="1">incorrect option</option>
            <option value="2">correct option</option>
          </>
        )}
      </select>
    </form>
  );
}

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);
```

**Comment by badsyntax** [comment, marked as answer]
Ok i figured it out, to use the Controller: https://codesandbox.io/s/react-hook-form-get-started-forked-ljkfu?file=/src/index.js:0-1137

```jsx
import React, { useEffect, useState } from "react";
import ReactDOM from "react-dom";
import { useForm, Controller } from "react-hook-form";

import "./styles.css";

function App() {
  const { control, handleSubmit, watch } = useForm({
    defaultValues: {
      example: "2"
    }
  });

  const onSubmit = (data) => {
    console.log(data);
  };

  // Change the default value to true to see the correct select state
  const [loaded, setLoaded] = useState(false);

  useEffect(() => {
    setTimeout(() => {
      setLoaded(true);
    }, 1000);
  }, []);

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <Controller
        control={control}
        name="example"
        render={({ field }) => {
          return (
            <select {...field}>
              {loaded && (
                <>
                  <option value="1">incorrect option</option>
                  <option value="2">correct option</option>
                </>
              )}
            </select>
          );
        }}
      />
    </form>
  );
}

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);

```

**Reply by bluebill1049** [reply]
Have you tried using the `reset` API?

**Reply by badsyntax** [reply]
For a simple form I could use the reset API, but I'm building a generic "form builder" where the options list can be generated from any async source so it's hard to track when to reset in a generic way, without tracking the loading state of many async operations. Not impossible, but a lot of work. Using the controller as demonstrated makes this all easier for me. 

I actually think the original issue I posted is a generic DOM or React issue as I was able to replicate without using react-hook-form.

Thanks for the awesome library btw. Its a very flexible library although it takes a while to learn the API and the right way to do things.

**Reply by bluebill1049** [reply]
thanks for the feedback.
